src/kOS/UserIO/TelnetWelcomeMenu.cs

```diff
@@ -25,12 +25,248 @@ public class TelnetWelcomeMenu : MonoBehaviour
         private readonly StringBuilder localMenuBuffer;
         private bool firstTime;
         private volatile bool forceMenuReprint;
-        
+        private MenuLevel currentLevel;
+
+        private abstract class MenuLevel
+        {
+            public virtual MenuLevel parentMenu { get { return new OuterMenu (); } }
+
+            public virtual List<String> helpText { get; } = new List<String> { };
+
+            public virtual List<List<String> > headers { get; } = new List<List<String> > {
+                new List<String> { "Menu", "" },
+                new List<String> { "Pick", "" },
+                new List<String> { "----", "----------------" }
+            };
+            public virtual String format { get; } = "{0,4} {1}";
+
+            public virtual void PrintMenu (TelnetWelcomeMenu parent)
+            {
+
+                List<String> displayChoices = new List<string> { };
+                int longestLength = 0;
+                foreach (List<String> hdrLine in headers) {
+                    String choice = String.Format (format, hdrLine.Cast<object> ().ToArray ());
+                    displayChoices.Add (choice);
+                    longestLength = Math.Max (choice.Length, longestLength);
+                }
+                int i = 1;
+                foreach (object[] bodyLine in this.MenuItems()) {
+                    string choice = String.Format (format, new object[] { "[" + i.ToString () + "]" }.Concat (bodyLine).ToArray ());
+                    displayChoices.Add (choice);
+                    longestLength = Math.Max (choice.Length, longestLength);
+                    i++;
+                }
+                foreach (string choice in displayChoices) {
+                    string choicePaddedToLongest = choice + new String (' ', (longestLength - choice.Length));
+                    parent.telnetServer.Write (parent.CenterPadded (choicePaddedToLongest, ' ') + (char)UnicodeCommand.STARTNEXTLINE);
+                }
+            }
+
+            public virtual List< object[] > MenuItems ()
+            { // Deliberately empty, so this is allowed not to be implemented.
+                return new List< object[] > ();
+            }
+
+            public virtual int ItemCount ()
+            {
+                return MenuItems ().Count;
+            }
+
+            public virtual bool ItemsChanged ()
+            {
+                return false;
+            }
+
+            public virtual void Selected (TelnetWelcomeMenu parent, int k)
+            {
+                return;
+            }
+        }
+
+        private class OuterMenu : MenuLevel
```

Comment:
"I need my neighbour to house-sit for me while I'm gone.  But because I want to uphold some weird esoteric principle, I refuse to do that by giving only him a copy of the key.  Instead I'll trust the entire world and just leave the house unlocked entirely."

I'm probably in the minority here, but I find friend classes less ugly than the alternative, which is to just open up a method to the whole public, rather than just limiting it to the classes you are in control of.

That being said, friend classes are ugly in that they are like Access Control Lists in file systems, where you have to explicitly whitelist individual users, when the cleaner solution is to have group permissions and allow individuals to be members of multiple groups at the same time.  But I'm not aware of any OOP languages that took that approach to the concept.